### PR TITLE
refactor(DMS): rabbitmq supports flaovr id

### PR DESCRIPTION
**Pull request by houpeng80**
**What this PR does / why we need it**:

- rabbitmq can be created by flavor id 
- rabbitmq support new format path of API

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note

```

## PR Checklist

* [ ] Tests added/passed.
* [ ] Documentation updated.
* [ ] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/dms/' TESTARGS='-run TestAccDmsRabbitmqInstances_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms/ -v -run TestAccDmsRabbitmqInstances_ -timeout 360m -parallel 4
=== RUN   TestAccDmsRabbitmqInstances_basic
=== PAUSE TestAccDmsRabbitmqInstances_basic
=== RUN   TestAccDmsRabbitmqInstances_newFormat_cluster
=== PAUSE TestAccDmsRabbitmqInstances_newFormat_cluster
=== RUN   TestAccDmsRabbitmqInstances_newFormat_single
=== PAUSE TestAccDmsRabbitmqInstances_newFormat_single
=== RUN   TestAccDmsRabbitmqInstances_withEpsId
=== PAUSE TestAccDmsRabbitmqInstances_withEpsId
=== RUN   TestAccDmsRabbitmqInstances_compatible
=== PAUSE TestAccDmsRabbitmqInstances_compatible
=== RUN   TestAccDmsRabbitmqInstances_single
=== PAUSE TestAccDmsRabbitmqInstances_single
=== CONT  TestAccDmsRabbitmqInstances_basic
=== CONT  TestAccDmsRabbitmqInstances_withEpsId
=== CONT  TestAccDmsRabbitmqInstances_single
=== CONT  TestAccDmsRabbitmqInstances_compatible
--- PASS: TestAccDmsRabbitmqInstances_single (471.93s)
=== CONT  TestAccDmsRabbitmqInstances_newFormat_single
--- PASS: TestAccDmsRabbitmqInstances_compatible (942.67s)
=== CONT  TestAccDmsRabbitmqInstances_newFormat_cluster
--- PASS: TestAccDmsRabbitmqInstances_withEpsId (1409.39s)
--- PASS: TestAccDmsRabbitmqInstances_basic (2962.02s)
--- PASS: TestAccDmsRabbitmqInstances_newFormat_single (2777.65s)
--- PASS: TestAccDmsRabbitmqInstances_newFormat_cluster (2863.78s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       3806.505s
```

### CODE REVIEW

File: go.sum [hunk 1]
```diff
@@ -26,6 +26,8 @@ github.com/census-instrumentation/opencensus-proto v0.2.1/go.mod h1:f6KPmirojxKA
 github.com/cespare/xxhash/v2 v2.1.1/go.mod h1:VGX0DQ3Q6kWi7AoAeZDth3/j3BFtOZR5XLFGgcrjCOs=
 github.com/chnsz/golangsdk v0.0.0-20230530080832-5259e6637c95 h1:inZTt5LVHKEUp/iVMvgQWiKeIZKB0n78Xn8uafU11YU=
 github.com/chnsz/golangsdk v0.0.0-20230530080832-5259e6637c95/go.mod h1:j6UR2TfACtmWBEvYrQqTpk5wy3b2QsEdiLkjMoM47j8=
+github.com/chnsz/golangsdk v0.0.0-20230531032846-271e0b51bd73 h1:0RCtUbvSK/IrTl4chRD7MHfHUsXn95J/xglVyagQA10=
+github.com/chnsz/golangsdk v0.0.0-20230531032846-271e0b51bd73/go.mod h1:j6UR2TfACtmWBEvYrQqTpk5wy3b2QsEdiLkjMoM47j8=
```
Comment (Lance52259) [inline go.sum:28]: go mod tidy pls

File: docs/resources/dms_rabbitmq_instance.md [hunk 2]
```diff
@@ -17,18 +17,47 @@ variable security_group_id {}
 
 data "huaweicloud_availability_zones" "zones" {}
 
-data "huaweicloud_dms_product" "test" {
-  engine        = "rabbitmq"
-  instance_type = "cluster"
-  version       = "3.7.17"
-  node_num      = 3
+data "huaweicloud_dms_rabbitmq_flavors" "test" {
+  type = "cluster"
```
Comment (chengxiangdong) [inline docs/resources/dms_rabbitmq_instance.md]:
```suggestion
data "huaweicloud_dms_rabbitmq_flavors" "test" {
  type              = "cluster"
  storage_spec_code = "dms.physical.storage.ultra.v2"
```

Comment (chengxiangdong) [reply docs/resources/dms_rabbitmq_instance.md]: It is recommended to specify the disk IO specification, and it is not recommended to use the default.

File: docs/resources/dms_rabbitmq_instance.md [hunk 3]
```diff
@@ -17,18 +17,47 @@ variable security_group_id {}
...
+  engine_version    = data.huaweicloud_dms_rabbitmq_flavors.test.versions[0]
+  storage_spec_code = data.huaweicloud_dms_rabbitmq_flavors.test.flavors[0].ios[0].storage_spec_code
+  broker_num        = 3
+
+  vpc_id             = var.vpc_id
+  network_id         = var.subnet_id
+  security_group_id  = var.security_group_id
+  availability_zones = [
+    data.huaweicloud_availability_zones.zones.names[0]
+  ]
```
Comment (chengxiangdong) [inline docs/resources/dms_rabbitmq_instance.md]: It is recommended to define a variable, such like:
````
variable "availability_zones" {
   default = ["your_availability_zones_a", "your_availability_zones_b", "your_availability_zones_c"]
}
```

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rabbitmq_instance_test.go [hunk 4]
```diff
@@ -72,6 +72,104 @@ func TestAccDmsRabbitmqInstances_basic(t *testing.T) {
...
+	rName := acceptance.RandomAccResourceNameWithDash()
+	updateName := acceptance.RandomAccResourceNameWithDash()
+	resourceName := "huaweicloud_dms_rabbitmq_instance.test"
+	rc := acceptance.InitResourceCheck(
+		resourceName,
+		&instance,
+		getDmsRabbitMqInstanceFunc,
+	)
+
+	// DMS instances use the tenant-level shared lock, the instances cannot be created or modified in parallel.
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rabbitmq_instance_test.go]: pls remove this comment as it is no longer needed

File: huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rabbitmq_instance_test.go [hunk 5]
```diff
@@ -72,6 +72,104 @@ func TestAccDmsRabbitmqInstances_basic(t *testing.T) {
...
```
Comment (chengxiangdong) [inline huaweicloud/services/acceptance/dms/resource_huaweicloud_dms_rabbitmq_instance_test.go]: ditto

File: docs/resources/dms_rabbitmq_instance.md [hunk 6]
```diff
@@ -17,18 +17,47 @@ variable security_group_id {}
...
+data "huaweicloud_dms_rabbitmq_flavors" "test" {
+  type = "cluster"
 }
 
 resource "huaweicloud_dms_rabbitmq_instance" "test" {
   name              = "instance_1"
-  product_id        = data.huaweicloud_dms_product.test.id
-  engine_version    = data.huaweicloud_dms_product.test.version
-  storage_spec_code = data.huaweicloud_dms_product.test.storage_spec_code
+  product_id        = data.huaweicloud_dms_rabbitmq_flavors.test.flavors[0].flavor.id
```
Comment (chengxiangdong) [inline docs/resources/dms_rabbitmq_instance.md]: `product_id` has deprecated, please use `flavor_id`

File: docs/resources/dms_rabbitmq_instance.md [hunk 7]
```diff
@@ -17,18 +17,47 @@ variable security_group_id {}
...
+
+data "huaweicloud_availability_zones" "zones" {}
+
+data "huaweicloud_dms_rabbitmq_flavors" "test" {
+  type = "single"
+}
+
+resource "huaweicloud_dms_rabbitmq_instance" "test" {
+  name              = "instance_1"
+  product_id        = data.huaweicloud_dms_rabbitmq_flavors.test.flavors[0].flavor.id
```
Comment (chengxiangdong) [inline docs/resources/dms_rabbitmq_instance.md]: ditto

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go [hunk 8]
```diff
@@ -247,6 +260,20 @@ func resourceDmsRabbitmqInstanceCreate(ctx context.Context, d *schema.ResourceDa
 	config.MutexKV.Lock(lockKey)
 	defer config.MutexKV.Unlock(lockKey)
```
Comment (chengxiangdong) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go]: It seems that there is no global lock now, it is recommended to try to remove it.

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go [hunk 9]
```diff
@@ -267,13 +294,92 @@ func resourceDmsRabbitmqInstanceCreate(ctx context.Context, d *schema.ResourceDa
...
+		MinTimeout:   10 * time.Second,
+		PollInterval: 15 * time.Second,
+	}
+
+	if _, err = stateConf.WaitForStateContext(ctx); err != nil {
+		return diag.Errorf("error waiting for RabbitMQ instance (%s) to be ready: %s", instanceID, err)
+	}
+
+	// Store the instance ID now
+	d.SetId(instanceID)
```
Comment (chengxiangdong) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go]: suggest to move this line after line:335

File: go.sum (hunk 1 above)
Comment (houpeng80) [reply go.sum:28]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go [hunk 10]
```diff
@@ -267,13 +294,92 @@ func resourceDmsRabbitmqInstanceCreate(ctx context.Context, d *schema.ResourceDa
...
+	log.Printf("[INFO] Creating RabbitMQ instance, ID: %s", instanceID)
+
+	stateConf := &resource.StateChangeConf{
+		Pending:      []string{"CREATING"},
+		Target:       []string{"RUNNING"},
+		Refresh:      rabbitmqInstanceStateRefreshFunc(client, instanceID),
+		Timeout:      d.Timeout(schema.TimeoutCreate),
+		Delay:        300 * time.Second,
+		MinTimeout:   10 * time.Second,
+		PollInterval: 15 * time.Second,
```
Comment (chengxiangdong) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go]:
```suggestion
		Delay:        300 * time.Second,
		PollInterval: 15 * time.Second,
```
PollInterval: override MinTimeout/backoff and only poll this often

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go [hunk 11]
```diff
@@ -267,13 +294,91 @@ func resourceDmsRabbitmqInstanceCreate(ctx context.Context, d *schema.ResourceDa
...
+
+func createRabbitMQInstanceWithProductID(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	cfg := meta.(*config.Config)
+	region := cfg.GetRegion(d)
+	client, err := cfg.DmsV2Client(region)
+	if err != nil {
+		return diag.Errorf("error initializing DMS RabbitMQ(v2) client: %s", err)
+	}
+	var availableZones []string // Available zones IDs
+	azIDs, ok := d.GetOk("available_zones")
```
Comment (Lance52259) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go]: the `availability_zones` is the recommand parameter, not `available_zones`.

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go [hunk 12]
```diff
@@ -473,7 +578,7 @@ func resourceDmsRabbitmqInstanceUpdate(ctx context.Context, d *schema.ResourceDa
 		}
 	}
 
-	if d.HasChange("product_id") {
+	if d.HasChanges("product_id", "flavor_id") {
 		err = resizeRabbitMQInstance(ctx, d, meta, engineRabbitMQ)
```
Comment (Lance52259) [inline huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go:582]: split this function to `resizeRabbitMQInstanceByFlavorId` and `resizeRabbitMQInstanceByProductId`?

**Comment by houpeng80** [comment]
make testacc TEST='./huaweicloud/services/acceptance/dms/' TESTARGS='-run TestAccDmsRabbitmqInstances_'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/dms/ -v -run TestAccDmsRabbitmqInstances_ -timeout 360m -parallel 4
=== RUN   TestAccDmsRabbitmqInstances_basic
=== PAUSE TestAccDmsRabbitmqInstances_basic
=== RUN   TestAccDmsRabbitmqInstances_newFormat_cluster
=== PAUSE TestAccDmsRabbitmqInstances_newFormat_cluster
=== RUN   TestAccDmsRabbitmqInstances_newFormat_single
=== PAUSE TestAccDmsRabbitmqInstances_newFormat_single
=== RUN   TestAccDmsRabbitmqInstances_withEpsId
=== PAUSE TestAccDmsRabbitmqInstances_withEpsId
=== RUN   TestAccDmsRabbitmqInstances_compatible
=== PAUSE TestAccDmsRabbitmqInstances_compatible
=== RUN   TestAccDmsRabbitmqInstances_single
=== PAUSE TestAccDmsRabbitmqInstances_single
=== CONT  TestAccDmsRabbitmqInstances_basic
=== CONT  TestAccDmsRabbitmqInstances_withEpsId
=== CONT  TestAccDmsRabbitmqInstances_single
=== CONT  TestAccDmsRabbitmqInstances_compatible
--- PASS: TestAccDmsRabbitmqInstances_single (453.72s)
=== CONT  TestAccDmsRabbitmqInstances_newFormat_single
--- PASS: TestAccDmsRabbitmqInstances_withEpsId (742.12s)
=== CONT  TestAccDmsRabbitmqInstances_newFormat_cluster
--- PASS: TestAccDmsRabbitmqInstances_compatible (868.32s)
--- PASS: TestAccDmsRabbitmqInstances_basic (975.66s)
--- PASS: TestAccDmsRabbitmqInstances_newFormat_single (900.08s)
--- PASS: TestAccDmsRabbitmqInstances_newFormat_cluster (1105.58s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/dms       1847.746s

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go (hunk 11 above)
Comment (houpeng80) [reply huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go]: done

File: huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go (hunk 12 above)
Comment (houpeng80) [reply huaweicloud/services/dms/resource_huaweicloud_dms_rabbitmq_instance.go:582]: they are both modify flavor, and the param of broker_num and storage_space will also support modification, so it is better to put them together

**Pull request merged**